File: src/lib/relays/orchestration/hydration-priority.ts
export type HydrationPriority =
  | 'visible'
  | 'near-visible'
  | 'active-offscreen'
  | 'hidden-paused'
  | 'background'
  | 'cancelled';

export type HydrationJob = {
  readonly key: string;
  readonly priority: HydrationPriority;
  readonly generation: number;
};

const priorityRank: Record<HydrationPriority, number> = {
  visible: 0,
  'near-visible': 1,
  'active-offscreen': 2,
  'hidden-paused': 3,
  background: 4,
  cancelled: 5,
};

export function planHydrationJobs(input: {
  readonly jobs: readonly HydrationJob[];
  readonly generation: number;
}): {
  readonly runnable: readonly HydrationJob[];
  readonly cancelled: readonly string[];
} {
  const cancelled: string[] = [];
  const byKey = new Map<string, HydrationJob>();
  for (const job of input.jobs) {
    if (job.generation !== input.generation || job.priority === 'cancelled') {
      cancelled.push(job.key);
      continue;
    }
    if (job.priority === 'hidden-paused') continue;
    const current = byKey.get(job.key);
    if (!current || priorityRank[job.priority] < priorityRank[current.priority])
      byKey.set(job.key, job);
  }
  return {
    runnable: [...byKey.values()].sort(compareHydrationJobs),
    cancelled,
  };
}

function compareHydrationJobs(left: HydrationJob, right: HydrationJob): number {
  const byPriority = priorityRank[left.priority] - priorityRank[right.priority];
  return byPriority || left.key.localeCompare(right.key);
}

File: crates/lkjstr-app/src/hydration_priority.rs
#![doc = "Visibility-prioritized hydration scheduler reducer."]

#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
pub enum HydrationPriority {
    Visible,
    NearVisible,
    ActiveOffscreen,
    HiddenPaused,
    Background,
    Cancelled,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HydrationJob {
    pub key: String,
    pub priority: HydrationPriority,
    pub generation: u64,
}

#[must_use]
pub fn plan_hydration_jobs(
    jobs: &[HydrationJob],
    active_generation: u64,
) -> (Vec<HydrationJob>, Vec<String>) {
    let mut cancelled = Vec::new();
    let mut by_key = std::collections::BTreeMap::<String, HydrationJob>::new();
    for job in jobs {
        if job.generation != active_generation || job.priority == HydrationPriority::Cancelled {
            cancelled.push(job.key.clone());
            continue;
        }
        if job.priority == HydrationPriority::HiddenPaused {
            continue;
        }
        by_key
            .entry(job.key.clone())
            .and_modify(|current| {
                if job.priority < current.priority {
                    *current = job.clone();
                }
            })
            .or_insert_with(|| job.clone());
    }
    let mut runnable = by_key.into_values().collect::<Vec<_>>();
    runnable.sort_by(|a, b| a.priority.cmp(&b.priority).then_with(|| a.key.cmp(&b.key)));
    (runnable, cancelled)
}
